lib/segment/src/vector_storage/async_io.rs

```diff
@@ -63,24 +63,36 @@ impl UringReader {
         Ok(Self {
             file,
             buffers,
-            io_uring,
+            io_uring: Some(io_uring),
             raw_size,
             header_size,
         })
     }
 
+    pub fn drop_io_uring(&mut self) {
+        self.io_uring = None;
+    }
+
+    fn ensure_io_uring(&mut self) -> OperationResult<()> {
+        if self.io_uring.is_none() {
+            self.io_uring = Some(IoUring::new(DISK_PARALLELISM as _)?);
+        }
```

Comment:
You _can_ use [`Option::get_or_insert_with`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with) here to remove the if statement, but this is also totally fine of course.

```suggestion
        self.io_uring.get_or_insert_with(|| IoUring::new(DISK_PARALLELISM as _)?);
```

The _get_ part is optimized away if it isn't used.

Reply:
Can't propagate an error with `get_or_insert_with`. :(

Reply:
Very good point. Totally missed that. Thanks!